Added Footprint Collision Checker (Simple Commander API)

afifswaidan opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3151 |
| Primary OS tested on | Ubuntu 22.04.01 |
| Robotic platform tested on | None |

---

## Description of contribution in a few bullet points
* Added Footprint Collision Checker (Simple Commander API)
* Added one missing function in costmap_2d.py, it was needed in the footprint collision checker


---


#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

afifswaidan commented:
@SteveMacenski This is the initial commit, I still need to add documentation, docstrings and test cases. But I wanted to take your opinion if there are any specific test cases that I should implement?
Thanks!

SteveMacenski commented on nav2_simple_commander/nav2_simple_commander/footprint_collision_checker.py:
```diff
@@ -0,0 +1,96 @@
+#! /usr/bin/env python3
+# Copyright 2021 Samsung Research America
```
Needs unit testing :smile:

SteveMacenski commented on nav2_simple_commander/nav2_simple_commander/footprint_collision_checker.py:
```diff
@@ -0,0 +1,96 @@
+from math import cos, sin
+from line_iterator import LineIterator
+from costmap_2d import PyCostmap2D
+from geometry_msgs.msg import Point
```
Instead, take in a geometry_msgs/Polygon (which has that as one of its fields essentially)

afifswaidan replied:
Will do that, thanks for the notice!

SteveMacenski commented on nav2_simple_commander/nav2_simple_commander/costmap_2d.py:
```diff
@@ -175,6 +175,26 @@ def worldToMap(self, wx: float, wy: float) -> tuple[int, int]:
         my = int((wy - self.origin_y) // self.resolution)
         return (mx, my)
 
+    def worldToMapCheck(self, wx: float, wy: float) -> bool:
```
Ah, if this is only checking, how about `isWorldCoordsValid`

SteveMacenski replied:
Though, you seem to use this right before attempting to get the coords anyway. Why not return both the bool for if valid + the coordinates if they are? That way its a single check to get validity + results

SteveMacenski replied:
Could be `worldToMapValidated` or something

afifswaidan replied:
I tried merging both into 1 function named it worldToMap that returns validity + results. But, I found it sometimes confusing to the end user as they don't know that it is used for validity directly from its name.
I like the `isWorldCoordsValid` name for a separate function only for validity.
What do you think?

SteveMacenski replied:
We have a version of this in C++ that is `worldToMapWithBounds` basically that enforces bounds. I think here it would be sensible to have a single call that is `worldToMapValidated` where you can return `None`'s if the coordinate requests are invalid. Then in the usage, just check if the return values are `None` if its valid. I think that would be the most "pythonic" way to handle it

afifswaidan replied:
Okay so if I understood correctly, you want to make them one function called `worldToMapValidated` where it returns either the map coordinates XY (if the requested coordinates are valid) or None (if the requested coordinates are invalid).
This sounds good to me, I'll do it and commit the soonest.
Thanks! :)

SteveMacenski replied:
Yup!

SteveMacenski commented on nav2_simple_commander/nav2_simple_commander/footprint_collision_checker.py:
```diff
@@ -0,0 +1,96 @@
+    def setCostmap(self, costmap: PyCostmap2D):
+        self.costmap_ = costmap
+
+    def footprintCostAtPose(self, x: float, y: float, theta: float, footprint: list[Point]):
```
As well with Polygon message

SteveMacenski replied:
Also, a doc string would be nice here to explain the units (do we assume the X,Y,Theta are in world or map coords? Is the footprint in world or map coords?

afifswaidan replied:
Alright noted,  I wanted to add the docstrings and the unit tests after I take your initial opinion.
I will replace the point message with the polygon message, thanks for the notice!

SteveMacenski commented on nav2_simple_commander/nav2_simple_commander/footprint_collision_checker.py:
```diff
@@ -0,0 +1,96 @@
+
+class FootprintCollisionChecker():
+    def __init__(self, costmap: PyCostmap2D):
+        self.costmap_ = costmap
```
Remove the costmap setting on constructor and only on `setCostmap`. The fact that the C++ version does both is an oddity due to history that we can start fresh from in Python

afifswaidan replied:
I agree with you, I found it was redundant but I was taking the C++ version as a reference, I prefer to remove it as well.
Will do that in the next commit tomorrow hopefully.

SteveMacenski commented on nav2_simple_commander/nav2_simple_commander/footprint_collision_checker.py:
```diff
@@ -0,0 +1,96 @@
+    def __init__(self, costmap: PyCostmap2D):
+        self.costmap_ = costmap
+
+    def footprintCost(self, footprint: list[Point]):
```
a doc string would be nice here to explain the units: Is the footprint in world or map coords?

afifswaidan replied:
Alright noted,  I wanted to add the docstrings and the unit tests after I take your initial opinion.

SteveMacenski reviewed (changes requested):
LGTM for a first run, just a little cleanup and should be able to merge in soon!

afifswaidan commented:
I added the doc strings, migrated to Polygon and Point32 messages, and renamed the validity function.
Once you check them let me know if I need to modify anything. After that i will implement the test cases and commit them.
Thanks! :)

SteveMacenski commented:
Looked it all over and resolved the comments solved + added more recommendations!

SteveMacenski commented:
any update?

afifswaidan commented:
> any update? I would like this in before Christmas to button up that ticket for 2022 :smile: 

I was sick for the past week so I couldn't work or contribute. I already implemented some test cases, I'll continue the rest and commit on the weekend as usual. Is that okay? 😊
I believe we can finish it before Christmas for sure.

SteveMacenski commented:
Awesome, totally understand. Feel better! I appreciate the help :-)

SteveMacenski reviewed (changes requested):
Otherwise looks good to me! Thanks for the unit tests.

afifswaidan commented on nav2_simple_commander/nav2_simple_commander/footprint_collision_checker.py:
```diff
@@ -0,0 +1,213 @@
+        x1 = 0.0
+        y1 = 0.0
+
+        if self.worldToMapValidated(footprint.points[0].x, footprint.points[0].y) == None:
```
I got your point, then the function will always return either [X,Y] as valid coordinates or [None,None] if invalid.
Thanks for the comment, I will commit asap this weekend!

afifswaidan commented:
> https://app.circleci.com/pipelines/github/ros-planning/navigation2/8434/workflows/9cbe962d-b231-4b47-a10b-212aefdd07af/jobs/28245
> 
> Your tests in CI are failing due to linting and a few tests - please also fix 😄

I just checked the results, minor issues that can be easily fixed!
I will commit asap in the weekend.

afifswaidan commented:
@SteveMacenski any idea why this stderr is being raised only on this function?
The docstrings of other functions follow the same scheme with no errors.
![image](https://user-images.githubusercontent.com/53655365/206906810-bb4927b4-61bf-4066-b45a-fc987554729a.png)

I will fix it as suggested in the image but I thought to ask you because it confused me a bit :)

SteveMacenski commented:
I've seen things like that before, look at other file's docstrings and copy their formatting. I eventually figured it out with some help of python developers so those files are right! I know, its confusing and poorly documented

afifswaidan commented:
> I've seen things like that before, look at other file's docstrings and copy their formatting. I eventually figured it out with some help of python developers so those files are right! I know, its confusing and poorly documented
> 
> Edit: it looks like you have `Args:` there instead of `Args` the other functions have

Originally it was `Args` just as other functions have. Then the error came out and the error message suggested to change it to `Args:` so when I changed it the error didn't appear again.

afifswaidan commented:
> Like I said, looking at the other files might help
> 
> * https://github.com/ros-planning/navigation2/blob/main/nav2_simple_commander/nav2_simple_commander/costmap_2d.py
> * https://github.com/ros-planning/navigation2/blob/main/nav2_simple_commander/nav2_simple_commander/robot_navigator.py
> 
> They all have a bunch of functions, all covered as examples. I notice that its the only function that doesn't return a value like this one (https://github.com/ros-planning/navigation2/blob/main/nav2_simple_commander/nav2_simple_commander/costmap_2d.py#L121-L133) might be worth trying adding that None return field and see if that resolves

Yes, I did the docstrings exactly based on other previously merged/committed files.
As I noticed, It should be `Args: ` when the docstring doesn't contain a `Returns ` section. And it should be `Args ` when the docstring contains a `Return ` section. I added 2 photos as a comparison.
I will add a return None for the sake of keeping the docstrings consistent as all of them are `Args ` and not `Args: `.

![image](https://user-images.githubusercontent.com/53655365/207429512-8612eed6-1352-421d-874c-b248d0834408.png)
![image](https://user-images.githubusercontent.com/53655365/207429583-1b0e5df5-7acf-4089-bb2e-cdd8dfa9fbf8.png)

SteveMacenski commented:
Awesome!

i think the last thing was the verified worldtomap None return for 1 call!

afifswaidan commented:
> Awesome!
> 
> i think the last thing was the verified worldtomap None return for 1 call!

This was already implemented a few commits back. Now the function returns None in case of invalid inputs, or returns the [X,Y] coordinates from world coordinates to map coordinates.

SteveMacenski commented:
We talked about changing that to `worldToMapValidated -> [None, None]` so that you only called the function 1 time and you check `x, y` for being none.

afifswaidan commented:
> We talked about changing that to `worldToMapValidated -> [None, None]` so that you only called the function 1 time and you check `x, y` for being none.  

Ah yes correct, I modified the function but still kept calling it more than one time.
My bad. It is an easy fix, I will fix it and commit ASAP. I'll try my best to do it tomorrow or at least before the weekend.
Thanks for pointing it out again.

SteveMacenski commented on nav2_simple_commander/test/test_footprint_collision_checker.py:
```diff
@@ -0,0 +1,139 @@
+LETHAL_OBSTACLE = 254
+
+
+class TestFootprintCollisionChecker(unittest.TestCase):
```
Last thing : the unit tests should have some that have non-zero costs in them to make sure things are working properly. If you make a 100x100 map with a 10x10 block of cost in the middle (or something) and then run these same tests to make sure we're picking things up correctly.

Then this is good to merge! Thanks!

SteveMacenski commented:
nav2_simple_commander.test.test_footprint_collision_checker.TestFootprintCollisionChecker failed

```
test/test_footprint_collision_checker.py:67: in test_worldToMapValidated
    self.assertEqual(fcc_.worldToMapValidated(0, 5), None)
E   AssertionError: (None, None) != None
```

Also a linting error

```
AssertionError: Found 1 code style errors / warnings:
  ./nav2_simple_commander/footprint_collision_checker.py:70:1: W293 blank line contains whitespace
assert 1 == 0
test/test_flake8.py:23: in test_flake8
    assert rc == 0, \
E   AssertionError: Found 1 code style errors / warnings:
E     ./nav2_simple_commander/footprint_collision_checker.py:70:1: W293 blank line contains whitespace
E   assert 1 == 0
```

SteveMacenski commented:
@afifswaidan I forgot - can you add docs https://navigation.ros.org/commander_api/index.html about this feature?

SteveMacenski commented:
https://discourse.ros.org/t/nav2-costmap-collision-checking-python-api-added-to-nav2-simple-commander/28766

afifswaidan commented:
> @afifswaidan I forgot - can you add docs https://navigation.ros.org/commander_api/index.html about this feature? 

Sure thing!
I have 2 question, I should add the documentation for this collision checker and for the line iterator that I did previously right?

My 2nd question is, its my first time of adding/modifying a docs website, how do you usually do it for ros2 nav2 docs? Sorry if its a basic question but I'd like to know how and do it!

Thanks :)

afifswaidan commented:
> https://discourse.ros.org/t/nav2-costmap-collision-checking-python-api-added-to-nav2-simple-commander/28766

Thank you so much for the credits!
I must say that contributing under your supervision/assistance was a pleasant and beneficial experience for me.

Please ping me whenever there are some pending issues that needs contributors, I've been working with ROS Navigation since 2017 and I hope I can be a good help for the development of ROS.

SteveMacenski commented:
https://github.com/ros-planning/navigation.ros.org/blob/master/commander_api/index.rst Basically just modify this file under the `Costmap API` section to mention the collision checker and provide a link to it

SteveMacenski commented:
Something that would be very helpful if you were interested is https://github.com/ros-planning/navigation2/issues/3205. The idea is that if you create a `while true` loop, you may notice that the controller can run much faster than if you throttle it down to a set rate. The project is to try to find a way so that we can leverage more of that speed. It might be timer we use isn't working as we'd like, context switching of threads, I mis-diagnosed the issue when  I was developing something else, etc.

afifswaidan commented:
> https://github.com/ros-planning/navigation.ros.org/blob/master/commander_api/index.rst Basically just modify this file under the `Costmap API` section to mention the collision checker and provide a link to it 

Okay sure I'll do that the soonest.
Thanks for the help!

afifswaidan commented:
> Something that would be very helpful if you were interested is https://github.com/ros-planning/navigation2/issues/3205. The idea is that if you create a `while true` loop, you may notice that the controller can run much faster than if you throttle it down to a set rate. The project is to try to find a way so that we can leverage more of that speed. It might be timer we use isn't working as we'd like, context switching of threads, I mis-diagnosed the issue when  I was developing something else, etc. 

Okay sure, I can look into it.
But can you give me some details about the timeline?
When is it expected to be finished?
Because next week I might not be fully available.

Thanks for telling me about it as well.

SteveMacenski commented:
I don't expect much to get done from tomorrow to the end of the year from anyone :laughing: 

When you have time and interest :-) If this doesn't interest you, there are plenty of things in the issue tracker and would very much appreciate contributions where you would have interest!

afifswaidan commented:
> I don't expect much to get done from tomorrow to the end of the year from anyone :laughing: 
> 
> When you have time and interest :-) If this doesn't interest you, there are plenty of things in the issue tracker and would very much appreciate contributions where you would have interest! 

Yeah exactly its the holidays and end lf year season :laughing: 

Okay I'll check it out and let you know. I'll check the other issues as well.

afifswaidan commented:
Hello Steve,
I wish you a Merry Christmas.

I will be looking into all the stuff you told me about later in the first week of January 2023.
I noticed something that I wanted to ask you about.
In our last commit, a wrong email address was included in the details of the commit, can this be fixed somehow?
My actual email address is afifswaidan.as@gmail.com.
Some other account is linked to the commit I don't actually know why.
Thanks in advance!

![image](https://user-images.githubusercontent.com/53655365/209571080-44ad8b99-5552-46f7-ae59-9ff6e14febf9.png)

SteveMacenski commented:
Unfortunately, that’s to do with your git configurations and not something I can control. Perhaps its possible in Git to manually manipulate that, but it would be wrong (and possibly problematic legally) for me to do that as a representative of a company. 

What we can do instead though is revert the commit so its out of the git history and you can resubmit a new PR under your preferred address. Its mostly problematic for me as a 3rd party to reassign the commit. For you, its fine if its more properly representative.

afifswaidan commented:
> Unfortunately, that’s to do with your git configurations and not something I can control. Perhaps its possible in Git to manually manipulate that, but it would be wrong (and possibly problematic legally) for me to do that as a representative of a company. 
> 
> What we can do instead though is revert the commit so its out of the git history and you can resubmit a new PR under your preferred address. Its mostly problematic for me as a 3rd party to reassign the commit. For you, its fine if its more properly representative. 

Thanks for for the quick reply.
Of course I don't want us to do anything that is potentially problematic.
So if I understood correctly, you revert back the repo, and I submit a new PR with corrected git configs?

afifswaidan commented:
> https://github.com/ros-planning/navigation.ros.org/blob/master/commander_api/index.rst Basically just modify this file under the `Costmap API` section to mention the collision checker and provide a link to it

I modified the file based on previous APIs (costmap and commander)
Let me know if anything else should be added.
